### DISCUSSION TITLE
How to simulate a travelling electron beam's spacial-temporal current profile

### DISCUSSION THREAD

**Discussion by ZhixinFan-hub** [svenreiche/Genesis-1.3-Version4#119] [Q&A]
Hi
What I want to simulate is an electron beam with a Gaussian density profile on the Z-axis traveling through the undulator. The code I tried is:

````
&setup
rootname=prxt
lattice=prxt.lat
beamline=FEL
lambda0=1e-10
gamma0=1135
delz=0.1
seed=1
shotnoise=true
one4one=true
write_semaphore_file=true
&end

&time
s0=0
slen=1e-8
sample=1
time=false
&end

&profile_gauss
label=currenttest
c0=3000
s0=0
sig=3e-9
&end

&lattice
zmatch=9.5
&end

&field
power=5e3
dgrid=2.000000e-04
ngrid=255
waist_size=30e-6
&end

&beam
current=@currenttest
delgam=1.000000
ex=4.000000e-07
ey=4.000000e-07
&end


&write
field=fld
beam=beam
&end

&track
&end
````

The lattice file is:

```
D1: DRIFT = { l = 0.44};
D2: DRIFT = { l = 0.24};
QF: QUADRUPOLE = { l = 0.080000, k1= 2.000000 };
QD: QUADRUPOLE = { l = 0.080000, k1= -2.000000 };
UND: UNDULATOR = { lambdau=0.015000, nwig=266, aw=0.84853, helical= True};

FODO: LINE={UND,D1,QF,D2,UND,D1,QD,D2};

FEL: LINE={2*FODO};

```

In my assumption, in both spatial and temporal domain, the current distribution should be a gaussain envelop whose center travels through the undulator as time goes by, but the result is different:

<img width="916" alt="image" src="https://github.com/svenreiche/Genesis-1.3-Version4/assets/65019083/b57e96d6-f7aa-41fd-987b-c4f779cd845d">

As is shown in the figure, in the time domain I get half of gaussain envelope, and in the spatial domain, it is a constant when the time is determined.

For the half of gaussain envelope in the time domain, I tried to set s0=1.5e-9, as is shown in the figure below:
<img width="1190" alt="image" src="https://github.com/svenreiche/Genesis-1.3-Version4/assets/65019083/dc3ba8f1-0167-4ef3-9cdc-aeb49aa38b31">
It proves that by adjusting s0, I can get a whole Gaussain envelope in the time domain, but how can I get that in the spatial domain?

**Comment by svenreiche** [comment]
Hi,

Note that the output in z is not a spatial dimension but the independent variable of the Hamiltonean after a Legendre transformation. If you want to express the current profile so it is in the 's' variable, as it is also used in the and profile_gauss namelists

Z is a global parameter( point of observation) since the current profile is the same when passing through an observation point at z=0 and z=2m , one will not see any Dependance with z.

Hope this makes it clesr.

Best
Sven

Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ZhixinFan-hub** [reply]
Hi
So where is the 's' variable....

**Comment by svenreiche** [comment]
S is the arrival time schlecht by the speed of light



Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by svenreiche** [comment]
Sorry  Geman autocorrect

S is the scaled arrival time (by c) at a given position z within the undulator


Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ZhixinFan-hub** [reply]
Hi
Sorry, I don't understand...
In profile_gauss name list, there is only 's0', not 's'

So how can I simulate a spatial gaussian current profile....

**Comment by svenreiche** [comment]
First define in the time namelist the range of s, which goes from s0 to s1=s0+Sven

Second, deine the gaussian profile with the profile_gauss name list.

The gauss is definitiv as c0 × exp(-(s-s0)^2/2 sigma^2)

Where the values of s are defined in the first step

That's it. Note that time and spatial distribution in the longitudinal direction are the same apart from the scaling factor 'c'



Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ZhixinFan-hub** [reply]
Hi
Can you add a time-dependent example to the manual?
There is no example in your manual, which makes it pretty hard to understand.

**Reply by ZhixinFan-hub** [reply]
I have tried your suggestions, it seems to work.
But, what I want is the current's spatial-temporal evolution, e.g. I=I(z,t), where the range of z and t is from the start (z=0,t=0) to the end(electrons leave the undulator).
But as you said: "s1=s0+Sven",  it seems that the length of z is limited to 7 meters...

**Comment by svenreiche** [comment]
I think you are mixing up s and z.
Why would you define a bunch length of 7 m.
S is the distance in the co-moving frame of the electron bunches and moves along the undulator.





-----------------------------------

Paul Scherrer Institut

Sven Reiche
WBGB/119
CH-5232 Villigen PSI

Telefon: +41 56 310 51 19
E-Mail: ***@***.******@***.***>

________________________________

**Reply by ZhixinFan-hub** [reply]
As you said here:
![image](https://github.com/svenreiche/Genesis-1.3-Version4/assets/65019083/cd25906b-0dc0-43bc-a617-e279f4185f41)

**Comment by svenreiche** [comment]
Sorry that was an auto correct.

It should be S0 to s0+slen

where slen is the input parameter in the time namelist

Sven


-----------------------------------

Paul Scherrer Institut

Sven Reiche
WBGB/119
CH-5232 Villigen PSI

Telefon: +41 56 310 51 19
E-Mail: ***@***.******@***.***>

________________________________

**Reply by ZhixinFan-hub** [reply]
So, in the output file, only the data from `s0` to `s1=s0+slen` are recorded. `s0=s0_{initial}+c*t`, where `s0_{initial}` is the 's0' value we set in `&time`, c is the speed of light and t is the time.

Am I correct?

**Comment by svenreiche** [comment]
Almost,

it is the co-moving frame of the electron bunch and not the radiation field. So the correct scaling is

s0 = s0_{initial} + c*<beta>*t

where <beta> is the average beta of the electron beam.





-----------------------------------

Paul Scherrer Institut

Sven Reiche
WBGB/119
CH-5232 Villigen PSI

Telefon: +41 56 310 51 19
E-Mail: ***@***.******@***.***>

________________________________

**Comment by ZhixinFan-hub** [comment]
OK, so in the `XXXX.out.h5` file, `/Beam/current` data was a `n*m` matrix, where `m` is how many slices it has, `m=slen/ lambda0/sample` where `lambda0` is in `&setup` and others are in`&time`, as is shown in the manual.
`n` is equal to the length of `/Lattice/zplot` data in the `XXXX.out.h5 file`.

In this figure, I label `s` as `time(a.u.)`, corresponding to the `m`, and the other axis as `position`, corresponding to the `n`.
![image](https://github.com/svenreiche/Genesis-1.3-Version4/assets/65019083/697dac61-d575-4cf5-8272-edc626bd972f)

As you said:s0 = s0_{initial} + c*<beta>*t, labeling `s` as `time(a.u.)` is roughly correct (it indicates the time).
So what about the `position`? I wonder the physical meaning of this dimension.

**Comment by svenreiche** [comment]
The definition of position is the position of a 'virtual detector' to measure the beam properties and when the particles arrives at the detector.

As I said before it is the independent variable of the Hamiton system

So in any entry  in /Lattice/zplotdata there is such detector analysing the beam


-----------------------------------

Paul Scherrer Institut

Sven Reiche
WBGB/119
CH-5232 Villigen PSI

Telefon: +41 56 310 51 19
E-Mail: ***@***.******@***.***>

________________________________

**Reply by ZhixinFan-hub** [reply]
So, for the two dimensions in `/Bea,/current`, one indicates where `s0` is, the other is the slices from `s0(t)` to `s1(t)=s0(t)+slen`.

Am I correct?

**Reply by lxdmay** [reply]
Does it mean that the electron beam current distribution corresponding to each Z position is the same? Just can't change it.

**Reply by svenreiche** [reply]
Genesis has only one rather advance option to allow a change in the current profile. This is by running one4one simulation and enabling 'sorting'. Otherwise the beam profile is filled one and a current value is assigned to a given slice in the electron bunch. This current profile is then the same along the undulator, which means at every z-position.
